File: packages/analytics/src/mock-sink.ts
import type {
  AnalyticsEvent,
  AnalyticsSink,
  SinkDeliverContext,
  SinkInitContext,
} from './types.js'

/** A mock sink that records every call — used for testing the router. */
export interface MockSink extends AnalyticsSink {
  /** All events received across all deliver() calls (flattened). */
  events: AnalyticsEvent[]
  /** Each deliver() call's batch + context. */
  deliveries: Array<{ batch: AnalyticsEvent[]; ctx: SinkDeliverContext }>
  /** init() call contexts. */
  initCalls: SinkInitContext[]
  /** flush() invocation count. */
  flushCalls: number
  /** shutdown() invocation count. */
  shutdownCalls: number
}

export interface CreateMockSinkOptions {
  name?: string
  consentCategories?: string[]
}

/**
 * createMockSink — an `AnalyticsSink` that captures everything for assertion.
 * Mirrors the testing ergonomics of `@refraction-ui/ai`'s mock providers.
 */
export function createMockSink(
  options: CreateMockSinkOptions = {},
): MockSink {
  const sink: MockSink = {
    name: options.name ?? 'mock',
    consentCategories: options.consentCategories,
    events: [],
    deliveries: [],
    initCalls: [],
    flushCalls: 0,
    shutdownCalls: 0,

    init(ctx: SinkInitContext): void {
      sink.initCalls.push(ctx)
    },

    deliver(batch: AnalyticsEvent[], ctx: SinkDeliverContext): void {
      sink.deliveries.push({ batch, ctx })
      for (const ev of batch) sink.events.push(ev)
    },

    flush(): void {
      sink.flushCalls++
    },

    shutdown(): void {
      sink.shutdownCalls++
    },
  }

  return sink
}

File: packages/flutter/lib/src/analytics/mock_sink.dart
/// A mock sink that records every call — used for testing the router and the
/// wire-contract conformance suite. Mirrors `createMockSink` from
/// `@refraction-ui/analytics`.
library;

import 'types.dart';

class MockDelivery {
  MockDelivery(this.batch, this.ctx);
  final List<AnalyticsEvent> batch;
  final SinkDeliverContext ctx;
}

class MockSink implements AnalyticsSink {
  MockSink({String? name, List<String>? consentCategories})
    : _name = name ?? 'mock',
      _consentCategories = consentCategories;

  final String _name;
  final List<String>? _consentCategories;

  /// All events received across all deliver() calls (flattened).
  final List<AnalyticsEvent> events = [];

  /// Each deliver() call's batch + context.
  final List<MockDelivery> deliveries = [];

  /// init() call contexts.
  final List<SinkInitContext> initCalls = [];

  /// flush() invocation count.
  int flushCalls = 0;

  /// shutdown() invocation count.
  int shutdownCalls = 0;

  @override
  String get name => _name;

  @override
  List<String>? get consentCategories => _consentCategories;

  @override
  void init(SinkInitContext ctx) {
    initCalls.add(ctx);
  }

  @override
  void deliver(List<AnalyticsEvent> batch, SinkDeliverContext ctx) {
    deliveries.add(MockDelivery(batch, ctx));
    events.addAll(batch);
  }

  @override
  void flush() {
    flushCalls++;
  }

  @override
  void shutdown() {
    shutdownCalls++;
  }
}

/// Create an [AnalyticsSink] that captures everything for assertion.
MockSink createMockSink({String? name, List<String>? consentCategories}) =>
    MockSink(name: name, consentCategories: consentCategories);
